### ISSUE TITLE
Racket attempts closing an extremely large amount of file descriptors when spawning a subprocess

### ISSUE THREAD

**Issue by whitequark** [racket/racket#5147]
**What version of Racket are you using?**
8.7 [cs]

**What program did you run?**

First:

```
raco pkg install rosette
```

Then:

```
#lang rosette/safe

(verify (assert #f))
```

The version of Rosette is:

```
User-specific for installation "8.7":
 Package  Checksum                Source
 rosette  cf703c60e2164fda5ea...  catalog...mina/rosette
```

**What should have happened?**

Rosette spawns Z3, reads the trivial unsat core, and returns an error.

**What actually happens**

Rosette hangs. If I strace the Racket process, I get a stream of `close(...) = EBADF` syscalls like:

```
strace: Process 1573258 attached
08:29:25.568928 close(1049178175)       = -1 EBADF (Bad file descriptor) <0.000007>
08:29:25.569018 close(1049178174)       = -1 EBADF (Bad file descriptor) <0.000006>
08:29:25.569045 close(1049178173)       = -1 EBADF (Bad file descriptor) <0.000007>
08:29:25.569070 close(1049178172)       = -1 EBADF (Bad file descriptor) <0.000008>
08:29:25.569095 close(1049178171)       = -1 EBADF (Bad file descriptor) <0.000005>
08:29:25.569112 close(1049178170)       = -1 EBADF (Bad file descriptor) <0.000005>
08:29:25.569128 close(1049178169)       = -1 EBADF (Bad file descriptor) <0.000005>
08:29:25.569148 close(1049178168)       = -1 EBADF (Bad file descriptor) <0.000006>
08:29:25.569171 close(1049178167)       = -1 EBADF (Bad file descriptor) <0.000006>
08:29:25.569194 close(1049178166)       = -1 EBADF (Bad file descriptor) <0.000005>
08:29:25.569211 close(1049178165)       = -1 EBADF (Bad file descriptor) <0.000004>
08:29:25.569226 close(1049178164)       = -1 EBADF (Bad file descriptor) <0.000006>
08:29:25.569259 close(1049178163)       = -1 EBADF (Bad file descriptor) <0.000007>
08:29:25.569282 close(1049178162)       = -1 EBADF (Bad file descriptor) <0.000007>
```

**Please include any other relevant details**

I've installed Racket from Debian.

I am reporting a bug against Racket even though I'm using Rosette because my experience with other language runtimes suggests that this is an issue with a Racket-wide subprocess spawn subroutine (but I don't know where specifically it's invoked).

The output of `ulimit -a -H` is as follows:

```
$ ulimit -a -H
real-time non-blocking time  (microseconds, -R) unlimited
core file size              (blocks, -c) unlimited
data seg size               (kbytes, -d) unlimited
scheduling priority                 (-e) 40
file size                   (blocks, -f) unlimited
pending signals                     (-i) 255639
max locked memory           (kbytes, -l) unlimited
max memory size             (kbytes, -m) unlimited
open files                          (-n) 1073741816
pipe size                (512 bytes, -p) 8
POSIX message queues         (bytes, -q) 819200
real-time priority                  (-r) 20
stack size                  (kbytes, -s) unlimited
cpu time                   (seconds, -t) unlimited
max user processes                  (-u) 255639
virtual memory              (kbytes, -v) unlimited
file locks                          (-x) unlimited
```

Racket is the only runtime I've used that exhibits this problem. I believe that old versions of Python 3 did as well, but they've since started using [close_range](https://man7.org/linux/man-pages/man2/close_range.2.html) to avoid spending O(n) time.

Even those old versions of Python limited the amount of files to be closed to something much more reasonable, I think 131072 or thereabouts? It was still slow enough to show in profiles, but it didn't make the application pin a core effectively indefinidely.

I do not know why the limit on the number of open files is set to such a high value. I have not done so myself (as far as I recall anyway) and `grep 1073741816 /etc -r` produces no results. Perhaps it is a new kernel or Debian default?

**Comment by whitequark** [comment]
The relevant Python issue is https://github.com/python/cpython/issues/86904.

**Comment by sorawee** [comment]
8.7 is more than 2 years ago. Could you try the 8.15, which is latest version (probably need to build from source for Debian) and see if there's still a problem?

**Comment by whitequark** [comment]
Thanks, looks fixed somewhere between 8.7 and 8.14.

For posterity, I found out that the 1073741816 rlimit comes from systemd.
